src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -455,94 +457,175 @@ static uct_iface_internal_ops_t uct_cuda_copy_iface_internal_ops = {
     .ep_is_connected       = uct_base_ep_is_connected
 };
 
-static UCS_CLASS_INIT_FUNC(uct_cuda_copy_iface_t, uct_md_h md, uct_worker_h worker,
-                           const uct_iface_params_t *params,
-                           const uct_iface_config_t *tl_config)
+void uct_cuda_copy_cleanup_per_ctx_rscs(uct_cuda_copy_per_ctx_rsc_t *ctx_rsc)
 {
-    uct_cuda_copy_iface_config_t *config = ucs_derived_of(tl_config,
-                                                          uct_cuda_copy_iface_config_t);
-    ucs_status_t status;
+    CUstream *stream;
+    ucs_queue_head_t *event_q;
     ucs_memory_type_t src, dst;
-    ucs_mpool_params_t mp_params;
+	unsigned long long ctx_id;
+    CUresult cu_err;
+    CUcontext popped_ctx;
 
-    UCS_CLASS_CALL_SUPER_INIT(uct_cuda_iface_t, &uct_cuda_copy_iface_ops,
-                              &uct_cuda_copy_iface_internal_ops, md, worker,
-                              params, tl_config, "cuda_copy");
+    cu_err = cuCtxGetId(ctx_rsc->cuda_ctx, &ctx_id);
+    /* check if context is still active */
+    if (cu_err == CUDA_SUCCESS) {
 
-    status = uct_cuda_base_check_device_name(params);
-    if (status != UCS_OK) {
-        return status;
+        cuCtxPushCurrent(ctx_rsc->cuda_ctx);
+        ucs_memory_type_for_each(src) {
+            ucs_memory_type_for_each(dst) {
+                stream  = &ctx_rsc->queue_desc[src][dst].stream;
+                event_q = &ctx_rsc->queue_desc[src][dst].event_queue;
+
+                if (!ucs_queue_is_empty(event_q)) {
+                    ucs_warn("stream destroyed but queue not empty");
+                }
+
+                if (*stream == 0) {
+                    continue;
+                }
+
+                UCT_CUDADRV_FUNC_LOG_ERR(cuStreamDestroy(*stream));
+            }
+        }
+
+        if (ctx_rsc->short_stream) {
+            UCT_CUDADRV_FUNC_LOG_ERR(cuStreamDestroy(ctx_rsc->short_stream));
+        }
+
+        ucs_mpool_cleanup(&ctx_rsc->cuda_event_desc, 1);
```

Comment:
Shouldn't this mpool cleanup be called even if the `ctx_rsc->cuda_ctx` is not valid anymore?

Reply:
If context is not present, then context could've been destroyed by the user before ucp_context_destroy or MPI_Finalize.